main.cpp

```diff
@@ -272,406 +221,213 @@ Use a service like https://www.diffchecker.com/diff to compare your output.
 
 #include <iostream>
 #include <cmath>
+#include <memory>
 #include <functional>
 
-struct DoubleType;
-struct IntType; 
-
-struct FloatType    
+template <typename NumericType>
+struct Numeric    
 {
-private:
-    float* value = nullptr;
-    FloatType& powInternal(const float ft);
-public:
-    FloatType(float fTValue) : value( new float (fTValue) ){}
-    ~FloatType()
-    {
-        delete value;
-        value = nullptr;
-    }
-
-    FloatType& operator+=(float rhs);
-    FloatType& operator-=(float rhs);
-    FloatType& operator*=(float rhs);
-    FloatType& operator/=(float rhs);
-
-    FloatType& pow(float ft);
-    FloatType& pow(const FloatType& ft);
-    FloatType& pow(const DoubleType& dt);
-    FloatType& pow(const IntType& it);
-    operator float() const { return *value; }
-
-    FloatType& apply(std::function<FloatType&(float&)> f);
-    FloatType& apply(void(*f)(float&));
-};
+    using Type = NumericType;
 
-//------------------------------------------------
+    Numeric(Type v) : value( std::make_unique<Type>(v) ){}
 
-struct DoubleType 
-{
-private:
-    double* value = nullptr;
-    DoubleType& powInternal(double dt);
-public:
-    DoubleType(double dtValue) : value( new double (dtValue) ){}
-    ~DoubleType()
+    ~Numeric()
```

Comment:
`unique_ptr` will manage the lifetime of the owned object.  you don't need to assign it to nullptr when your class goes out of scope.  This means that you don't even need a destructor.
- [ ] revise